qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,349 @@
+    on arbitrary quantum state preparation on page 3.
+
+    The only methods to be used externally are extractRomValuesFromState, angle2RomValue,
+    rotationTreeFromState and getAngle0.
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    @staticmethod
+    def extractRomValuesFromState(state: ArrayLike, rot_reg_size: int, adjoint: bool = False):
+        r"""Gives list in which the ith element is a list of the rom values to be loaded when
+        preparing the amplitudes of the ith qubit for the given state.
+        """
+        rotation_tree = RotationTree.rotationTreeFromState(state)
+        next_layer = [rotation_tree]
+        rom_vals = []
+        while len(next_layer) != 0:
+            this_layer = next_layer
+            next_layer = []
+            rom_vals_this_layer = []
+            for tree in this_layer:
+                angle = tree.getAngle0()
+                if adjoint:
+                    angle = 2 * np.pi - angle
+                rom_val = RotationTree.angle2RomValue(angle, rot_reg_size)
+                rom_vals_this_layer.append(rom_val)
+                if tree.branch0 is not None:
+                    next_layer.append(tree.branch0)
+                if tree.branch1 is not None:
+                    next_layer.append(tree.branch1)
+            rom_vals.append(rom_vals_this_layer)
+        return rom_vals
+
+    @staticmethod
+    def rotationTreeFromState(state):
+        r"""Given a list of coefficients, returns a tree-like object that contains the angles for
+        the rotations when preparing the state.
+        """
+        if len(state) == 2:
+            return RotationTree(abs(state[0]) ** 2, abs(state[1]) ** 2, None, None)
+        dn_l = RotationTree.rotationTreeFromState(state[: len(state) // 2])
+        dn_r = RotationTree.rotationTreeFromState(state[len(state) // 2 :])
+        return RotationTree(dn_l.sum_total, dn_r.sum_total, dn_l, dn_r)
+
+    def getAngle0(self):
+        r"""Get the angle that corresponds to p_0."""
+        return 2 * np.arccos(np.sqrt(self.__getP0()))
+
+    @staticmethod
+    def angle2RomValue(angle, rot_reg_size):
+        r"""Returns the value to be loaded to a QROM to encode the given angle with a certain value
+        of rot_reg_size.
+        """
+        rom_value_decimal = 2**rot_reg_size * angle / (2 * np.pi)
+        return round(rom_value_decimal) % (2**rot_reg_size)
+
+    # do not call, use rotationTreeFromState
+    def __init__(self, sum0, sum1, branch0, branch1):
+        self.sum0 = sum0
+        self.sum1 = sum1
+        self.sum_total = sum0 + sum1
+        self.branch0 = branch0
+        self.branch1 = branch1
+
+    def __getP0(self):
+        if self.sum_total == 0:
+            return 0
+        return self.sum0 / self.sum_total
```

Comment:
have a look in say and_bloq for building examples test cases which are good for testing the bloq protocols (`@bloq_example`).

You can also add to our autodocumentation tool: dev_tools/autogenerate-bloqs-notebooks-v2.py which helps document the bloqs separate from your nice tutorial. This is helpful to check your docs are rendering properly. This shouldn't be a blocker but something to think about.

Reply:
I tried adding my bloq to the autogen tool but it did nothing. Actually, I think that most of the bloqs there are not being generated, or at least it does not seem to be the case from the script's output. The commits are [aa82ece](https://github.com/quantumlib/Qualtran/pull/627/commits/aa82ece16af2e38ec512a03af25991960c836ba4) and [35e45c8](https://github.com/quantumlib/Qualtran/pull/627/commits/35e45c89877c6af45544b65fba755f5bc1b88fc7).

Reply:
Could you try the pattern from [here](https://github.com/quantumlib/Qualtran/blob/main/qualtran/bloqs/and_bloq.py#L213), i.e., I think something like

``` python

@bloq_example
def _controlled_state_prep():
    state_coefs = (
        (-0.42677669529663675 - 0.1767766952966366j),
        (0.17677669529663664 - 0.4267766952966367j),
        (0.17677669529663675 - 0.1767766952966368j),
        (0.07322330470336305 - 0.07322330470336309j),
        (0.4267766952966366 - 0.17677669529663692j),
        (0.42677669529663664 + 0.17677669529663675j),
        (0.0732233047033631 + 0.17677669529663678j),
        (-0.07322330470336308 - 0.17677669529663678j),
    )
    controlled_state_prep = ControlledStatePreparationUsingRotations(n_qubits=2, rot_reg_size=4, state=state_coefs)
    return controlled_state_prep
```
(the names have to match up to the _ which is stripped later on). Note to self: this should be documented better.

Reply:
As to why it's helpful, you can add [unit tests](https://github.com/quantumlib/Qualtran/blob/main/qualtran/bloqs/and_bloq_test.py#L30), which will run through several checks on your bloq which can be very helpful.

Reply:
Now it works! But the result looks a bit weird, because there is first the tutorial and then a giant block of documentation without any context. Maybe it would be a good idea to rename the tutorial as `controlled_state_preparation_tutorial.ipynb` and leave the autogen documentation as a separate file? This I think looks much cleaner.

Reply:
Okay so I figured out why the autogen doc wasn't working. It turns out that the command in the file that indicates how to run the script points to an older version. https://github.com/quantumlib/Qualtran/blob/061f17269ea8f06e46bf65bd2e81ab86e0e2bf15/dev_tools/autogenerate-bloqs-notebooks-v2.py#L41-L42
A `-v2` is missing from it. I copy-pasted it without checking and what was running was the old version. I fixed it in the most recent commit.

Reply:
right, we normally separate tutorials out from the bloq documentation if there is a detailed tutorial.